sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -436,57 +437,90 @@ static size_t _handle_req(coap_pkt_t *pdu, uint8_t *buf, size_t len,
     return pdu_len;
 }
 
+static int _request_matcher_default(gcoap_listener_t *listener,
+                                    const coap_resource_t **resource,
+                                    const coap_pkt_t *pdu)
+{
+    uint8_t uri[CONFIG_NANOCOAP_URI_MAX];
+    int ret = GCOAP_RESOURCE_MISMATCH;
+
+    if (coap_get_uri_path(pdu, uri) <= 0) {
+        return GCOAP_RESOURCE_NO_PATH;
+    }
+
+    coap_method_flags_t method_flag = coap_method2flag(coap_get_code_detail(
+                                                           (coap_pkt_t *)pdu));
+
+    for (size_t i = 0; i < listener->resources_len; i++) {
+        *resource = &listener->resources[i];
+
+        int res = coap_match_path(*resource, uri);
+
+        /* URI mismatch */
+        if (res > 0) {
+            continue;
+        }
+        /* resources expected in alphabetical order */
+        else if (res < 0) {
+            return GCOAP_RESOURCE_ERROR;
+        }
+
+        /* potential match, check for method */
+        if (! ((*resource)->methods & method_flag)) {
+            /* record wrong method error for next iteration, in case
+             * another resource with the same URI and correct method
+             * exists */
+            ret = GCOAP_RESOURCE_WRONG_METHOD;
+            continue;
+        }
+        else {
+            return GCOAP_RESOURCE_FOUND;
+        }
+    }
+
+    return ret;
+}
+
 /*
  * Searches listener registrations for the resource matching the path in a PDU.
  *
+ * param[in]  pdu -- the PDU to check the resource for
  * param[out] resource_ptr -- found resource
  * param[out] listener_ptr -- listener for found resource
  * return `GCOAP_RESOURCE_FOUND` if the resource was found,
  *        `GCOAP_RESOURCE_WRONG_METHOD` if a resource was found but the method
  *        code didn't match and `GCOAP_RESOURCE_NO_PATH` if no matching
```

Comment:
If this returns NO_PATH if there's just no match found, what is the MISMATCH then for?